models/mask_rcnn.py

```diff
@@ -0,0 +1,1273 @@
+  roi_bin_grid_w = sampling_ratio if exact_sampling else (roi_width / pooled_width).ceil()
+
+  if exact_sampling:
+    count = max(roi_bin_grid_h * roi_bin_grid_w, 1)  
+    iy = Tensor.arange(roi_bin_grid_h, device=input.device) 
+    ix = Tensor.arange(roi_bin_grid_w, device=input.device) 
+    ymask = None
+    xmask = None
+  else:
+    count = (roi_bin_grid_h * roi_bin_grid_w).maximum(1)
+    iy = Tensor.arange(height, device=input.device)  
+    ix = Tensor.arange(width, device=input.device)  
+    ymask = iy[None, :] < roi_bin_grid_h[:, None] 
+    xmask = ix[None, :] < roi_bin_grid_w[:, None] 
+
+  def from_K(t):
+    return t[:, None, None]
+
+  y = (
+    from_K(roi_start_h)
+    + ph[None, :, None] * from_K(bin_size_h)
+    + (iy[None, None, :] + 0.5) * from_K(bin_size_h / roi_bin_grid_h)
+  )
+  x = (
+    from_K(roi_start_w)
+    + pw[None, :, None] * from_K(bin_size_w)
+    + (ix[None, None, :] + 0.5) * from_K(bin_size_w / roi_bin_grid_w)
+  )
+
+  val = _bilinear_interpolate(input, roi_batch_ind, y, x, ymask, xmask)
+  if not exact_sampling:
+    val = ymask[:, None, None, None, :, None].where(val, 0)
+    val = xmask[:, None, None, None, None, :].where(val, 0)
+
+  output = val.sum((-1, -2))
+  if isinstance(count, Tensor):
+    output /= count[:, None, None, None]
+  else:
+    output /= count
+
+  output = output.cast(orig_dtype)
+  return output
+
+class ROIAlign:
+  def __init__(self, output_size, spatial_scale, sampling_ratio):
+    self.output_size = output_size
+    self.spatial_scale = spatial_scale
+    self.sampling_ratio = sampling_ratio
+
+  def __call__(self, input, rois):
+    output = _roi_align(
+      input, rois, self.spatial_scale, self.output_size[0], self.output_size[1], self.sampling_ratio, aligned=False
+    )
+    return output
+
+
+class LevelMapper:
+  def __init__(self, k_min, k_max, canonical_scale=224, canonical_level=4, eps=1e-6):
+    self.k_min = k_min
+    self.k_max = k_max
+    self.s0 = canonical_scale
+    self.lvl0 = canonical_level
+    self.eps = eps
+
+  def __call__(self, boxlists):
+    s = Tensor.sqrt(Tensor.cat(*[boxlist.area() for boxlist in boxlists]))
+    target_lvls = (self.lvl0 + Tensor.log2(s / self.s0 + self.eps)).floor()
+    target_lvls = target_lvls.clip(min_=self.k_min, max_=self.k_max)
+    return target_lvls - self.k_min
+
+
+class Pooler:
+  def __init__(self, output_size, scales, sampling_ratio):
+    self.output_size = output_size
+    self.scales = scales
+    self.sampling_ratio = sampling_ratio
+    poolers = []
+    for scale in scales:
+      poolers.append(
+        ROIAlign(
+          output_size, spatial_scale=scale, sampling_ratio=sampling_ratio
+        )
+      )
+    self.poolers = poolers
+    self.output_size = output_size
+    lvl_min = float(-Tensor.log2(Tensor([scales[0]], dtype=dtypes.float32)).numpy()[0])
+    lvl_max = float(-Tensor.log2(Tensor([scales[-1]], dtype=dtypes.float32)).numpy()[0])
```

Comment:
Is there a need to go through tinygrad and back here?

Reply:
no, this can be directly writen with math.log2

Reply:
changes this, though I expect negligible performance impact as this is not run many times during inference